pkg/securitycontextconstraints/seccomp/withseccomp.go

```diff
@@ -3,111 +3,174 @@ package seccomp
+		// This means that we now have to automatically allow `runtime/default`
+		// if a user specifies `docker/default` and vice versa in an SCC.
+		if p == v1.DeprecatedSeccompProfileDockerDefault || p == v1.SeccompProfileRuntimeDefault {
+			runtimeDefaultAllowed = true
+		}
+		allowed = append(allowed, p)
+	}
 
-// NewWithSeccompProfile creates a new must run as strategy or returns an error if it cannot
-// be created.
-func NewWithSeccompProfile(allowedProfiles []string) (SeccompStrategy, error) {
-	return &withSeccompProfile{allowedProfiles}, nil
+	return &strategy{
+		allowedProfiles:       allowed,
+		allowAnyProfile:       allowAny,
+		runtimeDefaultAllowed: runtimeDefaultAllowed,
+	}
 }
 
-// Generate creates the profile based on policy rules.
-func (s *withSeccompProfile) Generate(pod *api.Pod) (string, error) {
+// Generate returns a profile based on constraint rules.
+func (s *strategy) Generate(podAnnotations map[string]string, pod *api.Pod) (string, error) {
+	if podAnnotations[api.SeccompPodAnnotationKey] != "" {
+		// Profile already set, nothing to do.
+		return podAnnotations[api.SeccompPodAnnotationKey], nil
+	}
+	if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
+		// Profile field already set, translate to annotation
+		return podutil.SeccompAnnotationForField(pod.Spec.SecurityContext.SeccompProfile), nil
+	}
+
 	// return the first non-wildcard profile
-	for _, p := range s.allowedProfiles {
-		if p != allowAnyProfile {
-			return p, nil
-		}
+	if len(s.allowedProfiles) > 0 {
+		return s.allowedProfiles[0], nil
 	}
-	// if we reached this point then either there are no allowed profiles (empty slice)
-	// or the only thing in the slice is the wildcard.  In either case just return empty
-	// which means use the runtime default.
+
 	return "", nil
 }
 
 // ValidatePod ensures that the specified values on the pod fall within the range
 // of the strategy.
-func (s *withSeccompProfile) ValidatePod(pod *api.Pod) field.ErrorList {
+func (s *strategy) ValidatePod(pod *api.Pod) field.ErrorList {
 	allErrs := field.ErrorList{}
-	fieldPath := field.NewPath("pod", "metadata", "annotations", api.SeccompPodAnnotationKey)
-
-	podProfile, _ := pod.Annotations[api.SeccompPodAnnotationKey]
-
-	if len(s.allowedProfiles) == 0 && podProfile != "" {
+	podSpecFieldPath := field.NewPath("pod", "metadata", "annotations").Key(api.SeccompPodAnnotationKey)
+	podProfile := pod.Annotations[api.SeccompPodAnnotationKey]
+	// if the annotation is not set, see if the field is set and derive the corresponding annotation value
+	if len(podProfile) == 0 && pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
+		podProfile = podutil.SeccompAnnotationForField(pod.Spec.SecurityContext.SeccompProfile)
+	}
 
-		allErrs = append(allErrs, field.Forbidden(fieldPath, "seccomp may not be set"))
+	if !s.allowAnyProfile && len(s.allowedProfiles) == 0 && podProfile != "" {
```

Comment:
i was wondering if this and the `!profileAllowed` condition below is different from the container case, it's not.

i'd like to see this logic in one place, shared by both `ValidatePod` and `ValidateContainer`, the boolean logic is important.